lsp/server/src/server.ts

```diff
@@ -33,93 +33,136 @@ const documents: TextDocuments<TextDocument> = new TextDocuments(TextDocument);
-		if (folders) {
-			folders.forEach((folder) => {
-				dockerContainer(folder);
-			});
-		}
-	});
-	documents.all().forEach((doc)=>validateTextDocument(doc));
-});
+}
 
-class SA4UConfig {
-	compilationDir: string
-	constructor(compilationDir: string) {
-		this.compilationDir = '/src/' + compilationDir;
+async function stopDockerContainer(): Promise<void> {
+	let container_ids;
+	try {
+		container_ids = (await execAsync(`docker container ls -q -f name=sa4u_z3_server`)).stdout.split('\n').join(' ');
+		await execAsync(`docker container stop ${container_ids}`);
+	} catch (err) {
+		console.log(err);
 	}
 }
 
-// TODO: it would be nice to automatically discover the compile_commands.json file.
-const readFileAsync = promisify(readFile);
-
-async function getSA4UConfig(projectRootDir: string) {
+async function getSA4UConfig(): Promise<SA4UConfig> {
+	let readConfig: string;
 	try {
-		const configJSON = JSON.parse((await readFileAsync(projectRootDir + "/" + ".sa4u.json")).toString());
-		return new SA4UConfig(configJSON['CompilationDir']);
+		readConfig = (await promises.readFile('.sa4u.json')).toString();
 	} catch (err) {
 		console.warn('unable to read SA4U config: ' + err);
 		return new SA4UConfig('');
 	}
+	let configJSON;
+	try {
+		configJSON = JSON.parse(readConfig);
+	} catch (err) {
+		console.warn(`Failed to parse .sa4u.json as a JSON: ${err}`);
+		return new SA4UConfig('');
+	}
+	if (configJSON['CompilationDir'] === 'undefined') {
+		return new SA4UConfig('');
+	}
+	if (!existsSync('.sa4u')){
+		try {
+			await promises.mkdir('.sa4u');
+		} catch (err) {
+			console.warn(`Failed to create .sa4u directory: ${err}`);
+			return new SA4UConfig(configJSON['CompilationDir']);
+		}
+	}
+	const directory = resolve('./');
+	let readCompileCommands;
+	try {
+		readCompileCommands = (await promises.readFile(configJSON['CompilationDir']+'/compile_commands.json')).toString().replace(directory, '/src/');
+	} catch (err) {
+		console.warn(`Failed to read ${configJSON['CompilationDir']}/compile_commands.json: ${err}`);
+		return new SA4UConfig('');
+	}
+	try {
+		await promises.writeFile('.sa4u/compile_commands.json', readCompileCommands);
+		return new SA4UConfig('.sa4u/');
+	} catch (err) {
+		console.warn(`Failed to write compile_commands.json into .sa4u directory: ${err}`);
+		return new SA4UConfig(configJSON['CompilationDir']);
+	}
 }
 
-async function dockerContainer(folder: WorkspaceFolder) {
-	let path = decodeURIComponent(folder.uri);
-	const filePath = path;
-	path = path.replace(/(^\w+:|^)\/\//, '');
-	path = path.replace(/:/, '');
+async function getIgnoredFiles(): Promise<SA4UIgnoredFiles> {
+	let readConfig: string;
+	try {
+		readConfig = (await promises.readFile('.sa4u.json')).toString();
+	} catch (err) {
+		console.warn('unable to read SA4U config: ' + err);
+		return new SA4UIgnoredFiles([]);
+	}
+	let configJSON;
+	try {
+		configJSON = JSON.parse(readConfig);
+		if (configJSON['IgnoreFiles'] === 'undefined') {
```

Comment:
This should just be `if (configJSON['IgnoreFiles'] === undefined)` (no single quotes around undefined)

Reply:
Fixed